### DISCUSSION TITLE
How to properly update workspace when a dsl file is renamed ?

### DISCUSSION THREAD

**Discussion by ydaveluy** [eclipse-langium/langium#1666] [Q&A] [answered]
Hello,

I have a file `a.mydsl` that defines types and a second file `b.mydsl` with references to types defined in `a.mydsl`. If I rename `a.mydsl` to an extension not handled by my language such as `a.mydsl.old`, the file `b.mydsl` remains valid.
I have to restart VSCode to have invalid references in `b.mydsl`.

It seems that Langium keeps a document in the index in case the file extension is changed and not supported by the language.
Deletion works well, if I remove `a.mydsl`, `b.mydsl` is correctly updated.

**Comment by ydaveluy** [comment]
When a.mydsl is renamed to a.mydsl.old, it calls first DefaultDocumentUpdateHandler.didChangeContent() then DefaultDocumentUpdateHandler.fireDocumentUpdate().

I think that the incorrect behavior comes from DefaultServiceRegistry.hasServices() that returns true for uri `a.mydsl.old`. 

Maybe that an other interface such as ServiceRegistry.hasValidFileExtension(uri:URI) would be more appropriate ?

**Comment by msujew** [comment, marked as answer]
Hey @ydaveluy,

This is intended behavior (at least in VS Code). See also https://github.com/eclipse-langium/langium/pull/1492 which contains links to a discussion surrounding this issue. In the end, we only listen to what the language client tells us. 

In VS Code, if you rename a file extension to something that doesn't have a language associated with it, it will keep the old language association. You will see that the renamed file keeps the same language id (in the status bar). We correctly identify that by taking a look at the text document provided by the language client.

You can disable that behavior by overriding the `ServiceRegistry`. I wouldn't recommend it though, because this already works exactly how it's supposed to work.

**Comment by ydaveluy** [comment]
Ok thanks for the recommendation :-)
